libbfddp/bfddp_extra.c

```diff
@@ -0,0 +1,858 @@
+	msg.header.length = htons(sizeof(struct bfddp_message_header)
+				  + sizeof(struct bfddp_echo));
+	msg.header.type = htons(ECHO_REPLY);
+
+	/* Payload data. */
+	gettimeofday(&tv, NULL);
+	msg.data.echo.dp_time =
+		hu64tonu64((uint64_t)((tv.tv_sec * 1000000) + tv.tv_usec));
+	msg.data.echo.bfdd_time = bfdd_time;
+
+	return bfddp_write_enqueue(bctx, &msg);
+}
+
+size_t
+bfddp_send_session_state_change(const struct bfd_session *bs)
+{
+	struct bfddp_message msg = {};
+
+	/* Prepare header. */
+	msg.header.version = BFD_DP_VERSION;
+	msg.header.length = htons(sizeof(msg.header) + sizeof(msg.data.state));
+	msg.header.type = htons(BFD_STATE_CHANGE);
+
+	/* Prepare payload. */
+	msg.data.state.lid = htonl(bs->bs_lid);
+	msg.data.state.rid = htonl(bs->bs_rid);
+	msg.data.state.state = (uint8_t)bs->bs_state;
+	msg.data.state.diagnostics = (uint8_t)bs->bs_rdiag;
+	msg.data.state.detection_multiplier = bs->bs_rdmultiplier;
+	msg.data.state.desired_tx = htonl(bs->bs_rtx);
+	msg.data.state.required_rx = htonl(bs->bs_rrx);
+	msg.data.state.required_echo_rx = htonl(bs->bs_rerx);
+
+	if (bs->bs_rcbit)
+		msg.data.state.remote_flags |= RBIT_CPI;
+	if (bs->bs_rdemand)
+		msg.data.state.remote_flags |= RBIT_DEMAND;
+
+	msg.data.state.remote_flags = htonl(msg.data.state.remote_flags);
+
+	return bfddp_write_enqueue(bs->bs_bctx, &msg);
+}
+
+size_t
+bfddp_session_reply_counters(struct bfddp_ctx *bctx, uint16_t id,
+			     const struct bfd_session *bs)
+{
+	struct bfddp_message rmsg = {};
+	uint16_t msglen =
+		sizeof(rmsg.header) + sizeof(rmsg.data.session_counters);
+
+	/* Fill in message header. */
+	rmsg.header.version = BFD_DP_VERSION;
+	rmsg.header.length = htons(msglen);
+	rmsg.header.type = htons(BFD_SESSION_COUNTERS);
+	rmsg.header.id = id;
+
+	/* Failed to find session. */
+	if (bs == NULL) {
+		/* Send answer anyway so it doesn't wait forever. */
+		return bfddp_write_enqueue(bctx, &rmsg);
+	}
+
+	/* Fill payload. */
+	rmsg.data.session_counters.control_input_bytes =
+		hu64tonu64(bs->bs_crx_bytes);
+	rmsg.data.session_counters.control_input_packets =
+		hu64tonu64(bs->bs_crx_packets);
+	rmsg.data.session_counters.control_output_bytes =
+		hu64tonu64(bs->bs_ctx_bytes);
+	rmsg.data.session_counters.control_output_packets =
+		hu64tonu64(bs->bs_ctx_packets);
+
+	return bfddp_write_enqueue(bs->bs_bctx, &rmsg);
+}
+
+/*
+ * Misc functions.
+ */
+uint64_t
+hu64tonu64(uint64_t value)
+{
+	union {
+		uint32_t v32[2];
+		uint64_t v64;
+	} vu, *vp;
+
+	vp = (void *)&value;
+	vu.v32[0] = htonl(vp->v32[1]);
+	vu.v32[1] = htonl(vp->v32[0]);
+
+	return vu.v64;
+}
+
+uint64_t
+nu64tohu64(uint64_t value)
+{
+	union {
+		uint32_t v32[2];
+		uint64_t v64;
+	} vu, *vp;
+
+	vp = (void *)&value;
+	vu.v32[0] = ntohl(vp->v32[1]);
+	vu.v32[1] = ntohl(vp->v32[0]);
+
+	return vu.v64;
+}
+
+static uint32_t
+apply_jitter(uint32_t total, bool dm_one)
+{
+	uint32_t jitter;
+
+	if (dm_one)
+		jitter = (uint32_t)(random() % 11);
+	else
+		jitter = (uint32_t)(random() % 26);
```

Comment:
Minor:

```
jitter = (uint32_t)(random() % (dm_one? 11 : 26));
```

Also consider using macros for `11`, `26` to know what these "magic" numbers mean

Reply:
I'll fix this.